docs/tutorials/google_earth_engine/part_2/clip_by_asset.rst

```diff
@@ -1,646 +1,648 @@
+        
+        # Create earth engine directory, will raise exception if it already exists
+        try:
+            ee.batch.data.createAsset({
+                'type': 'Folder',
+                'name': earth_engine_root_dir
+            })
+        except EEException as e:
+            if 'Cannot overwrite asset' not in str(e):
+                raise e
+
+        # Create user directory, will raise exception if it already exists
+        try:
+            ee.batch.data.createAsset({
+                'type': 'Folder',
+                'name': user_root_dir
+            })
+        except EEException as e:
+            if 'Cannot overwrite asset' not in str(e):
+                raise e
+
+        return user_root_dir
+
+
+2. The ``get_user_boundary_path`` function determines the path to the boundary asset for a given user. Create a new ``get_user_boundary_path`` function in :file:`gee/methods.py` with the following contents:
+
+.. code-block:: python
+
+    def get_user_boundary_path(user):
+        """
+        Get a unique path for the user boundary asset.
+
+        Args:
+            user (django.contrib.auth.User): the request user.
+
+        Returns:
+            str: the unique path for the user boundary asset.
+        """
+        user_asset_dir = get_asset_dir_for_user(user)
+        user_boundary_asset_path = os.path.join(user_asset_dir, 'boundary')
+        return user_boundary_asset_path
+
+3. Update the ``upload_shapefile_to_gee`` function in :file:`gee/methods.py` to call the new ``get_user_boundary_path`` function and then export the ``ee.FeatureCollection`` to an asset at that path: (no try/except)
+
+.. code-block:: python
+    :emphasize-lines: 29-39
+
+    def upload_shapefile_to_gee(user, shp_file):
+        """
+        Upload a shapefile to Google Earth Engine as an asset.
+
+        Args:
+            user (django.contrib.auth.User): the request user.
+            shp_file (shapefile.Reader): A shapefile reader object.
+        """
+        features = []
+        fields = shp_file.fields[1:]
+        field_names = [field[0] for field in fields]
+
+        # Convert Shapefile to ee.Features
+        for record in shp_file.shapeRecords():
+            # First convert to geojson
+            attributes = dict(zip(field_names, record.record))
+            geojson_geom = record.shape.__geo_interface__
+            geojson_feature = {
+                'type': 'Feature',
+                'geometry': geojson_geom,
+                'properties': attributes
+            }
+
+            # Create ee.Feature from geojson (this is the Upload, b/c ee.Feature is a server object)
+            features.append(ee.Feature(geojson_feature))
+
+        feature_collection = ee.FeatureCollection(features)
+
+        # Get unique folder for each user to story boundary asset
+        user_boundary_asset_path = get_user_boundary_path(user)
+        print(user_boundary_asset_path)

```

Comment:
Remove this print statement unless you intended for it to be part of the tutorial.